### PR TITLE
Add missing deprecation information to problem

### PR DESCRIPTION
**Pull request by reinsch82**
Fixes 
 - gradle/gradle-private#4084

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Review by donat — APPROVED** [review]

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java [hunk 1]
```diff
@@ -86,31 +85,37 @@ public void featureUsed(final DeprecatedFeatureUsage usage) {
...
+    private void reportDeprecation(final DeprecatedFeatureUsage usage, final ProblemDiagnostics diagnostics) {
+        InternalProblemReporter reporter = ((InternalProblems) problemsService).getInternalReporter();
+        Problem problem = reporter.create(new Action<InternalProblemSpec>() {
+            @Override
+            public void execute(InternalProblemSpec builder) {
+                InternalProblemSpec problemSpec = builder
+                    // usage.getKind() could be be part of the problem ID, however it provides hints on the problem provenance which should be modeled differently, maybe as location data.
+                    .id("deprecated-feature-used", "Deprecated feature used", GradleCoreProblemGroup.deprecation())
+                    .contextualLabel(usage.getSummary())
+                    .details(usage.getRemovalDetails())
```
[the remark this answers is not in the document]
Comment (reinsch82) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:102]: yes

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java [hunk 2]
```diff
@@ -86,31 +85,37 @@ public void featureUsed(final DeprecatedFeatureUsage usage) {
...
 
-    private static void addPossibleLocation(ProblemDiagnostics diagnostics, ProblemSpec genericDeprecation) {
+    private void reportDeprecation(final DeprecatedFeatureUsage usage, final ProblemDiagnostics diagnostics) {
+        InternalProblemReporter reporter = ((InternalProblems) problemsService).getInternalReporter();
+        Problem problem = reporter.create(new Action<InternalProblemSpec>() {
+            @Override
+            public void execute(InternalProblemSpec builder) {
+                InternalProblemSpec problemSpec = builder
+                    // usage.getKind() could be be part of the problem ID, however it provides hints on the problem provenance which should be modeled differently, maybe as location data.
+                    .id("deprecated-feature-used", "Deprecated feature used", GradleCoreProblemGroup.deprecation())
```
[the remark this answers is not in the document]
Comment (donat) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:100]: I'm fine with that. So essentially the deprecated feature's name (e.g. the method name) should be part of the ID, right?

[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:100]: Yes, my understanding is that the ID would distinguish between the different deprecations types.

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java [hunk 3]
```diff
@@ -86,31 +85,37 @@ public void featureUsed(final DeprecatedFeatureUsage usage) {
...
+                    .contextualLabel(usage.getSummary())
+                    .details(usage.getRemovalDetails())
+                    .documentedAt(usage.getDocumentationUrl())
+                    .additionalData("type", usage.getType().name());
```
[the remark this answers is not in the document]
Comment (reinsch82) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java]: Are you thinking of the whole stacktrace or just the file location we can figure out from the stacktrace?

[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java]: If we know a file, that would be a file location. Here I think we should have a new type of location: A stacktrace location. That would also indicate that for this problem the stack trace is actually interesting. You could combine this with attaching the problem to a throwable and say in the method that the stack trace of the throwable is important.

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationMessage.java [hunk 4]
```diff
@@ -28,18 +28,27 @@ class DeprecationMessage {
     private final String context;
     private final DocLink documentation;
     private final DeprecatedFeatureUsage.Type usageType;
-
-    DeprecationMessage(String summary, String removalDetails, @Nullable String advice, @Nullable String context, @Nullable DocLink documentation, DeprecatedFeatureUsage.Type usageType) {
+    private final String problemId;
+
+    DeprecationMessage(String summary,
```
Comment (wolfs) [inline platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationMessage.java:33]:
```suggestion
    DeprecationMessage(
        String summary,
```

💅  Please put the first argument on a new line for better diffability in the future.

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java [hunk 5]
```diff
@@ -86,31 +85,40 @@ public void featureUsed(final DeprecatedFeatureUsage usage) {
...
 
-    private static void addPossibleLocation(ProblemDiagnostics diagnostics, ProblemSpec genericDeprecation) {
+    private void reportDeprecation(final DeprecatedFeatureUsage usage, final ProblemDiagnostics diagnostics) {
+        InternalProblemReporter reporter = ((InternalProblems) problemsService).getInternalReporter();
+        Problem problem = reporter.create(new Action<InternalProblemSpec>() {
+            @Override
+            public void execute(InternalProblemSpec builder) {
+                InternalProblemSpec problemSpec = builder
+                    // usage.getKind() could be be part of the problem ID, however it provides hints on the problem provenance which should be modeled differently, maybe as location data.
+                    .id(usage.getProblemId(), usage.getSummary())
```
Comment (wolfs) [inline platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:100]: ❌ I don't think the summary is part of the ID. I think we need to introduce something new that is not the summary, since the summary is very context specific. We don't necessarily do this as part of this PR, though we should do it very soon.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy [hunk 6]
```diff
@@ -80,6 +80,9 @@ class KnownProblemIds {
         // categories from integration tests
         'TEST_PROBLEM',
         'generic:deprecation:plugin',
+        'generic:configuration-consumption-detachedconfiguration',
+        'generic:behaviour-using-task-validateplugins-without-applying-the-java-toolchain-plugin',
+        'generic:method-repositoryhandler-jcenter',
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy:85]: 🤔 I think we should now assert the deprecation problem in addition to the deprecation when we test the deprecation. Again, maybe not in this PR, but very soon.

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java (hunk 5 above)
Comment (reinsch82) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:100]: My assumption is that the summary is about as context specific as the id, since the id holds e.g. the class name and/or method name that is deprecated.
The elaborate description would be in "details".

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationMessageBuilder.java [hunk 7]
```diff
@@ -491,6 +515,7 @@ public static class DeprecateBehaviour extends DeprecationMessageBuilder<Depreca
 
         public DeprecateBehaviour(String behaviour) {
             this.behaviour = behaviour;
+            this.problemId = createDeprecationGroup("behaviour", behaviour);
```
[the remark this answers is not in the document]
Comment (reinsch82) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationMessageBuilder.java:518]: I'm not sure I'm getting it right overall.
We added the deprecated method name also to the id. That is also not context independent as I understand it.
What am I missing.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy (hunk 6 above)
Comment (reinsch82) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/problems/KnownProblemIds.groovy:85]: We plan to do that separately.
Already in the backlog.

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationMessageBuilder.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationMessageBuilder.java:518]: The method that has been deprecated is not context (aka usage) sensitive: We deprecated a method, and that will be the same method no matter where we are using that from.
What is context sensitive is where we use it. For example, we have the behaviour `Build service 'KotlinToolingDiagnosticsCollector_2118021439' is being used by task ':build-logic:binary-compatibility:checkKotlinGradlePluginConfigurationErrors' without the corresponding declaration via 'Task#usesService'`. This behaviour is very context (aka usage) specific.

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Review by donat — APPROVED** [review]

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationMessageBuilder.java [hunk 8]
```diff
@@ -343,6 +381,13 @@ String formatAdvice(List<String> replacements) {
         }
     }
 
+    public static String createDefaultDeprecationIdString(String... ids) {
+        return screamingSnakeToKebabCase(StringUtils.join(ids, "-"))
+            .replaceAll("[^a-z-]", "-")
+            .replaceAll("-+", "-")
+            .replaceAll("^-+|-+$", "");
```
Comment (donat) [inline platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationMessageBuilder.java:388]: I think we need some description here what is the expected input/output here and/or why we need regex here.

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (reinsch82) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java]: I'd like to do the stack as a follow up
https://github.com/orgs/gradle/projects/19/views/12?pane=issue&itemId=57852123

File: platforms/extensibility/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidatePlugins.java [hunk 9]
```diff
@@ -94,6 +94,7 @@ public void validateTaskClasses() throws IOException {
                     spec.getForkOptions().setExecutable(getLauncher().get().getExecutablePath());
                 } else {
                     DeprecationLogger.deprecateBehaviour("Using task ValidatePlugins without applying the Java Toolchain plugin.")
+                        .withProblemIdDisplayName("Using task ValidatePlugins without applying the Java Toolchain plugin.")
```
Comment (donat) [inline platforms/extensibility/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidatePlugins.java:97]: Consider overriding `withProblemIdDisplayName` with `withProblemIdDisplayName(String id, String displayName)`

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java [hunk 10]
```diff
@@ -171,7 +171,7 @@ public InternalProblemBuilder details(String details) {
     }
 
     @Override
-    public InternalProblemBuilder documentedAt(DocLink doc) {
+    public InternalProblemBuilder documentedAt(@Nullable DocLink doc) {
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java:174]: Unrelated change. Should be done in a dedicated PR.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemDefinition.java [hunk 11]
```diff
@@ -31,7 +31,7 @@ public class DefaultProblemDefinition implements Serializable, ProblemDefinition
     private final Severity severity;
     private final DocLink documentationLink;
 
-    DefaultProblemDefinition(
+    public DefaultProblemDefinition(
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemDefinition.java:34]: @VisibleForTesting

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblems.java [hunk 12]
```diff
@@ -26,17 +27,22 @@
 @ServiceScope(Scope.BuildTree.class)
 public class DefaultProblems implements InternalProblems {
 
+    private CurrentBuildOperationRef currentBuildOperationRef;
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblems.java:30]: ❌ 
```suggestion
    private final CurrentBuildOperationRef currentBuildOperationRef;
```

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/InternalProblemSpec.java [hunk 13]
```diff
@@ -45,7 +47,7 @@ public interface InternalProblemSpec extends ProblemSpec {
      *
      * @return this
      */
-    InternalProblemSpec documentedAt(DocLink doc);
+    InternalProblemSpec documentedAt(@Nullable DocLink doc);
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/InternalProblemSpec.java:50]: Same as above.

**Review by donat — CHANGES_REQUESTED** [review]

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Review by hegyibalint — APPROVED** [review]
LGTM on the technical side.

**Comment by reinsch82** [comment]
@bot-gradle test RFN
